File: matlab/bench.m
function bench()
% Add CORA to path
addpath(genpath('cora'));
% Add fig2svg to path
addpath(genpath('fig2svg'));

% Intialize pseudo random number generator for repeatable results:
rng(31415);
% Run micro benchmark:

% Intervals:
t_scalar_int = [];
t_matrix_int = [];
ns = 2:10:200;
for n=ns
    t_m = [];
    disp([mfilename '>> Running matrix operations for n = ' num2str(n)]);
    for k_run=1:100
        x = rand(n,1) + rand(n,1) .* interval(-ones(n,1), ones(n,1));
        P = rand(n,n) + rand(n,n) .* interval(-ones(n,n), ones(n,n));
        [t_scalar_int(k_run), t_m(k_run)] = runbench(x, P);
    end
    t_matrix_int = [t_matrix_int t_m];
end
save('benchresults.mat');
end

% Runs the actual benchmark code and returns execution times in microseconds:
function [t_scalar, t_matrix] = runbench(x, P)
N = 100;
tic;
for k=1:N
    res = sin(x(1));
    res = cos(x(1));
    res = x(1)^2;
    res = sqrt(x(1));
end
t_scalar = (1e6*toc)/N;
% Run matrix computations:
tic;
for k=1:N
    res = x'*P*x;
end
t_matrix = (1e6*toc)/N;
end

File: julia/bench.jl
using LinearAlgebra
using IntervalArithmetic
using BenchmarkTools
using Random
using Plots
using JLD2, FileIO

cd(dirname(@__FILE__()))

# Runs the actual benchmark code and returns execution times in microseconds:
function runbench(x, P)
    texec = []
    N = 100
    t0 = time_ns()
    for k in 1:N
        res = sin(x[1])
        res = cos(x[1])
        res = x[1]^2
        res = sqrt(x[1])
    end
    push!(texec, 1e-3*(time_ns()-t0)/N)
    # Run matrix computations:
    t0 = time_ns()
    for k in 1:N
        res = x'*P*x
    end
    push!(texec, 1e-3*(time_ns()-t0)/N)
end

# Intialize pseudo random number generator:
Random.seed!(31415);
# Run micro benchmark:

# Intervals:
t_scalar = [];
NN = 100;
let t_matrix = Array{Float64}(undef, NN, 0);
    ns = 2:10:200;
    for n in ns
        t_m = [];
        print("Running matrix operations for n = ")
        println(n)
        for k_run in 1:NN
            x = rand(n,1) + rand(n,1) .* map(Interval, -ones(n,1), ones(n,1))
            P = rand(n,n) + rand(n,n) .* map(Interval, -ones(n,n), ones(n,n))
            texec = runbench(x, P)
            push!(t_scalar, texec[1])
            push!(t_m, texec[2])
        end
        t_m = vec(t_m);
        t_matrix = hcat(t_matrix, t_m);
        println(size(t_matrix))
    end
    @save "results.jld2" t_scalar t_matrix ns
    println("Results saved to file")
end
